packages/jest-runtime/src/index.ts

```diff
@@ -1269,9 +1305,17 @@ class Runtime {
     from: InitialModule,
     options?: InternalModuleOptions,
   ): LocalModuleRequire {
-    // TODO: somehow avoid having to type the arguments - they should come from `NodeRequire/LocalModuleRequire.resolve`
-    const resolve = (moduleName: string, options: ResolveOptions) =>
-      this._requireResolve(from.filename, moduleName, options);
+    const resolve = (moduleName: string, resolveOptions?: ResolveOptions) => {
+      const resolved = this._requireResolve(
+        from.filename,
+        moduleName,
+        resolveOptions,
+      );
+      if (resolveOptions?.outsideJestVm && options?.isInternalModule) {
```

Comment:
thoughts on throwing if `outsideJestVm` but not internal?

Reply:
Nope nope nope it needs to be possible to load `expect`, `jest-diff` etc. inside the sandbox no matter what 🙃

Reply:
sure, but they wouldn't have `outsideJestVm`, right?

Reply:
If we want to optimize by taking `jest-diff` from outside in `expect` as you suggested (remains to be seen if that makes sense) they will - but should still be runnable inside

Reply:
Basically, I can very well imagine jest modules using this flag that may be loaded both inside and outside, and should receive an outside dependency either way

Reply:
I recommend throwing if it isn't internal for now and changing things later. Otherwise you are opening the floodgates now.

Reply:
To be clear about the implication, this would mean that a Jest package that we add a usage of this to might start throwing when a user requires it on their own in test code/dependencies, including things like `jest-worker` which are known to be used by Gatsby etc.
And I'm not sure our tests would detect it before publish, because there are some packages (like `jest-worker`) that we probably never require inside a sandbox.

Reply:
I'm not sure if I remember correctly but I think in the past when I used to work on this, internal modules were marked by a special flag when creating the VM context that was disabled later when user code was run. Re-using that to see if it should allow or disallow these requires should work for this use case, right?

Reply:
Yeah, the `isInternal` flag is only set by internal stuff, I don't see how  externals doing `require` is an issue?

Reply:
> I'm not sure if I remember correctly but I think in the past when I used to work on this, internal modules were marked by a special flag when creating the VM context that was disabled later when user code was run. Re-using that to see if it should allow or disallow these requires should work for this use case, right?

Correct, this is the `options?.isInternalModule` we're looking at - all of the new behavior is only applied when loading internal modules.
But this thread is about when you load one of Jest's modules using this mechanism from user code.

Having looked at our package structure a bit, this is a very realistic possibility ... let me give an example concerning the micromatch perf problems we had - that's a dependency of `jest-message-util`, which is a dependency of `expect`. If we e.g. make `jest-message-util` require outside `micromatch`, but then someone imports `expect` in their code (which some people definitely do because they don't like the global or use it as an invariant function in production code), `jest-message-util` will try to make an outside require but is not in an internal context, so their tests will break.

Reply:
It works when going via `runtime.requireInternalModule`, then that flag is set and the `require` injected into the vm is always internal. When we load user code, we do `runtime.requireModule` which does not set the `isInternal` flag.

We need to make sure that we inject a special internal `require.resolve` as well, only if `isInternal`. Unless I'm missing something it shouldn't affect anybody else.

Reply:
>but then someone imports expect in their code (which some people definitely do because they don't like the global or use it as an invariant function in production code)

why is this an issue? If they import `expect` all of that will be external, there's mismatch is there?

Reply:
We already have the test: https://github.com/facebook/jest/pull/9907/files#diff-09b0c97a052cb8720f6ceadf0772ce4dR58

At the moment, it asserts that the flag is ignored in an external context.
If we change the behavior as suggested, the test will need to assert that it throws with the flag in an external context.
But then that (throwing) is exactly what will happen if you load such a module in an external context (e.g. user code).
Not now of course, since we never use this flag, but the first time we add it somewhere (e.g. in `jest-message-util`)

Reply:
What about inverting the solution: the things you know will be pulled in from Jest could be required by the vm-creator and then put onto a special object on Jest, like `jest.internalModuleRegistry.require('micromatch')` which just does `return jest.$$internalModuleRegistry[moduleName]`.

That way you are only exposing a finite set of modules ahead of time instead of enabling ways to break out of the sandbox.

Reply:
Interesting alternative. The only obvious behavior difference that strikes me would be that it's per module (node), not per import (edge), meaning that if we declare `prettier` to be from outside, it'll be the outside prettier for ALL internal Jest modules. Could be good to avoid inconsistencies in using outside require vs sandbox require, or bad if we need the outside version in some places but the inside version in others. Also feels more 'magic' to me because you don't see anything special at the import site. Thoughts @simenb?

Reply:
ah, we're gonna add `require.resolve('', {outsideJestVm: true})`, never mind my examples.

Not sure a whitelist is better, then `runtime` would have to keep a list of all modules and what dependencies they'd need ahead of time instead if letting those modules decide themselves

Reply:
After thinking about it a bit, I think I also still prefer the solution in this PR, mostly for its explicitness (and because as Simen mentioned there's no whitelist that can have stale entries or such)